DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/README.md

```diff
@@ -0,0 +1,286 @@
+
+### Naive Kernel
+
+The desired memory access pattern in `NaiveKernel` is illustrated in the following diagram:
+
+![Naive Kernel Memory Access Pattern](./assets/access_pattern_naive.gif)
+
+In the `NaiveKernel`, the first array dimension (row) is accessed in an unrolled loop. In this way, the kernel tries to access all five elements in one column of the buffer, namely `array2d[0][col]` to `array2d[4][col]`. As a result, there are 4 loads and 5 stores to the memory system in each loop iteration, and since the initiation interval is 1, they **should** happen in a single clock cycle. The  [kernel memory viewer report](https://www.intel.com/content/www/us/en/docs/oneapi-fpga-add-on/developer-guide/current/kernel-memory-viewer.html) shows that the memory system for `array2d` has 8 banks. This report shows the stallable arbitration nodes that prevent the 4 loads and 5 stores from happening in a single clock cycle, and explains why the compiler decided to insert them.
+
+![Stallable Arbitration Node](./assets/stallable_arbitration_node.png)
+
+#### Making sense of the bank-selection bits
+
+The *Details* pane of the Kernel Memory Viewer report shows that the bank bits are the second to the fourth least significant bits, $b_4$, $b_3$, $b_2$. This means that integers in every other column will be in a different memory bank.
+
+The following table shows how each address encodes its bank of the memory system. Note how the banks do not align with the rows.
+
+|                 | Address | Bank bits | Bank # |
+|-----------------|---------|-----------|--------|
+| array2d[0][0]   | 0x0000  | 0b000     | 0      |
+| array2d[0][1]   | 0x0004  | 0b001     | 1      |
+| array2d[0][2]   | 0x0008  | 0b010     | 2      |
+| ...             |         |           |        |
+| array2d[0][7]   | 0x001C  | 0b111     | 7      |
+| array2d[0][8]   | 0x0020  | 0b000     | 0      |
+| array2d[0][9]   | 0x0024  | 0b001     | 1      |
+| array2d[0][10]  | 0x0028  | 0b010     | 2      |
+| ...             |         |           |        |
+| array2d[0][15]  | 0x003C  | 0b111     | 7      |
+| ...             |         |           |        |
+
+Similarly, the following table shows which memory bank each logical array element gets assigned to. Observe that each access to one column of `array2d` accesses more than one memory bank. Since more than one load and more than one store store is scheduled to a particular memory bank in the same clock cycle, the compiler inserts arbitration logic along with stallable load-store units (LSUs). You can see this in the Kernel Memory Viewer in the optimization report.
```

Comment:
```suggestion
Similarly, the following table shows which memory bank each logical array element gets assigned to. Observe that each access to one column of `array2d` accesses more than one memory bank. Since more than one load and more than one store store is scheduled to a particular memory bank in the same clock cycle, the compiler inserts arbitration logic along with stallable arbitration logic, which was shown above in the FPGA optimization report.
```